### PR TITLE
[MoE][NVIDIA] Add shared expert fusion for TRTLLM FP8

### PR DESCRIPTION
**Pull request by WoosukKwon**
## Summary

- Extend the FusedMoE abstraction so shared experts can be represented as always-on routed expert slots on NVIDIA GPUs.
- Enable the path for the NVIDIA DeepSeek-V3.2/GLM-5.2 model when routed and shared expert storage formats match.
- Run fused FP8 shared experts through the modular FlashInfer TRTLLM backend, including expert-row padding required by the kernel.
- Append and pack constant shared-expert routes in one Triton kernel, avoiding the extra top-k concatenation that regressed single-token decode.
- Preserve ROCm AITER's native fused-shared-expert routing behavior.

## Motivation

Shared-expert fusion was effectively limited to ROCm/AITER. NVIDIA models therefore launched the routed MoE and shared MLP separately even when both used compatible FP8 storage.

FlashInfer's native shared-expert entry point is not usable for this FP8 configuration, but its externally routed TRTLLM FP8 kernel handles the shared expert correctly when the shared rows and always-on route are included in the routed problem. This change adds that representation to FusedMoE and keeps the single-token path efficient by synthesizing the shared route while packing top-k IDs and weights.

## Duplicate-work check

I searched open PRs for `shared expert fusion`, `FusedMoE shared experts NVIDIA`, and `FlashInfer shared expert`, and inspected the related issue and PRs.

- #51695 standardizes model-level FSE enablement/quantization compatibility; it does not add NVIDIA routing, weight preparation, or FlashInfer execution. This PR may need a small rebase if that selection refactor lands first.
- #50963 tracks FlashInfer's native FP4 shared-expert API. This PR implements the externally routed TRTLLM **FP8** path and does not close that FP4 issue.
- #51632 fixes ROCm Triton expert-count alignment and does not implement the NVIDIA path.

No open PR found in those searches implements this change.

## Correctness

Real-weight model evaluation:

- Model: GLM-5.2-FP8 under `models/deepseek_v32/`
- Hardware: TP8 across two NVIDIA GB200 nodes
- Model override: `vllm.models.deepseek_v32.nvidia.model:DeepseekV32ForCausalLM`, forcing the NVIDIA implementation and avoiding the HF torch-compile path
- GSM8K exact match: **0.9325246**

Local tests on NVIDIA GB200:

```text
.venv/bin/python -m pytest tests/kernels/moe/test_routing.py -q
507 passed

.venv/bin/python -m pytest tests/models/deepseek_v32/test_sequence_parallel.py   tests/kernels/moe/test_routing.py tests/quantization/test_fp8.py   -k 'shared_expert_fusion or standard_router or trtllm_fp8_moe_validates_kernel_expert_count or flashinfer_trtllm_pads_fused_shared_expert_rows or trtllm_fp8_modular_uses_padded_fused_expert_count or trtllm_pack_appends_fused_shared_experts or monolithic_kernel_must_handle_fused_shared_experts_natively or fp8_moe_modular_apply_includes_fused_shared_experts' -q
14 passed, 524 deselected

.venv/bin/python -m pytest tests/kernels/moe/test_unquantized_backend_selection.py -q
26 passed, 1 skipped

.venv/bin/python -m pytest tests/kernels/moe/test_moe_layer.py   -k fused_shared_experts_cuda -q
1 passed, 510 deselected

.venv/bin/python -m pytest tests/models/deepseek_v32/test_sequence_parallel.py -q
4 passed

pre-commit run --files <all 28 changed files>
Passed, including ruff, mypy, typos, SPDX, and repository checks
```

## Performance

Dummy-weight GLM-5.2-FP8, TP8 across two 4xGB200 nodes, 16 input and 16 output tokens per request. The benchmark used the NVIDIA model override, `VLLM_USE_BREAKABLE_CUDAGRAPH=1`, `FULL_AND_PIECEWISE`, the default 83 regular capture sizes, prefix caching disabled, five warmup waves, and nine measured closed-loop burst waves per concurrency. Breakable graphs set compilation mode to `NONE`, so these results do not use torch/Inductor compilation.

| Concurrency | TTFT unfused | TTFT fused | Change | TPOT unfused | TPOT fused | Change |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 29.738 ms | 28.739 ms | -3.36% | 8.215 ms | 8.274 ms | +0.72% |
| 8 | 30.477 ms | 29.515 ms | -3.16% | 9.160 ms | 8.828 ms | -3.63% |
| 32 | 67.968 ms | 66.573 ms | -2.05% | 10.826 ms | 10.549 ms | -2.55% |
| 128 | 261.856 ms | 255.693 ms | -2.35% | 15.830 ms | 15.946 ms | +0.74% |
| 512 | 284.948 ms | 282.755 ms | -0.77% | 27.052 ms | 26.588 ms | -1.71% |

| Concurrency | Output throughput unfused | Output throughput fused | Change |
|---:|---:|---:|---:|
| 1 | 103.85 tok/s | 103.90 tok/s | +0.06% |
| 8 | 757.21 tok/s | 781.11 tok/s | +3.16% |
| 32 | 1,956.06 tok/s | 1,997.35 tok/s | +2.11% |
| 128 | 3,829.45 tok/s | 3,873.43 tok/s | +1.15% |
| 512 | 11,139.04 tok/s | 11,072.50 tok/s | -0.60% |

TTFT is queued-to-first-token time; median queue time remained below 0.6 ms. Fusion is neutral at concurrency 1 and improves median throughput by 1.1-3.2% at concurrency 8-128.

The fused concurrency-512 run was bimodal: five waves completed in 0.71-0.74 seconds and four in 0.90-0.92 seconds. Its p99 TTFT was 485.07 ms versus 306.78 ms unfused, and p99 TPOT was 39.48 ms versus 27.07 ms unfused. The median throughput result therefore does not show a benefit at concurrency 512, and the tail behavior needs further investigation.

### Kernel profile

I also captured matched PyTorch/CUPTI traces for fixed decode batches. These runs used the same dummy model, TP8 topology, `max_num_seqs=1024`, `max_num_batched_tokens=32768`, breakable `FULL_AND_PIECEWISE` CUDA graphs, and all 83 regular capture sizes as the end-to-end benchmark. The table reports the median steady-state GPU annotation on rank 0; profiler wall time is not used.

| Decode batch | Unfused main stream | Unfused shared stream | Fused main stream | Main-stream change |
|---:|---:|---:|---:|---:|
| 1 | 8.147 ms | 7.640 ms | 8.287 ms | +0.140 ms (+1.72%) |
| 8 | 9.145 ms | 8.622 ms | 8.811 ms | -0.334 ms (-3.65%) |
| 128 | 15.662 ms | 14.876 ms | 15.835 ms | +0.173 ms (+1.11%) |

The unfused shared-expert stream completes before the main stream at all three sizes, so its work is already hidden rather than added to the critical path. Fusion removes that competing stream and helps batch 8, but its modular routing overhead slightly exceeds the overlap benefit at batches 1 and 128. This matches the end-to-end TPOT trend.

The combined append-and-pack kernel costs 0.104-0.118 ms per decode step across 75 MoE layers (1.39-1.58 microseconds per layer). It has already removed the prior materialize-plus-concatenate path. Eliminating the remaining kernel would require a FlashInfer interface that can consume an implicit always-on shared route; no further safe vLLM-only optimization was identified from these traces.

### MoE operator microbenchmark

I separately measured one GLM-5.2-FP8 MoE layer from precomputed router
logits through routing, expert execution, runtime activation quantization, and
combine. The gate projection, TP communication/final all-reduce, attention,
scheduler, and frontend are excluded. Dummy BF16 weights were converted to
block FP8 before timing, so offline weight quantization is also excluded.

The benchmark uses the TP8 **per-rank** matrix shape on one GB200:
hidden size 6144, intermediate size 256 (= 2048 / TP8), 256 routed experts,
one shared expert, top-k 8, BF16 activations, block-FP8 [128, 128] weights,
and GLM-5.2 sigmoid/noaux routing with routed scale 2.5. It configures
`tp_size=1` with pre-sharded weights, so it is a kernel microbenchmark rather
than an actual distributed TP8 run.

Every token count has its own manually captured exact-shape
`torch.cuda.CUDAGraph`; vLLM model-runner CUDA graph dispatch is disabled.
The main baseline uses the monolithic FlashInfer TRTLLM routed MoE plus a
separate block-FP8 DeepGEMM shared MLP. To give the baseline its strongest
overlap behavior, `VLLM_SHARED_EXPERTS_STREAM_TOKEN_THRESHOLD=32768` forces
the shared MLP onto its auxiliary stream at every measured size; instrumentation
confirmed `MULTI_STREAM_OVERLAPPED` at both 1 and 8192 tokens. The fused path
uses the modular FlashInfer TRTLLM kernel with 257 logical experts padded to
260. Results are the median of three process-level medians on the same GB200.

| Tokens | Fused latency | Main always-overlap latency | Fused change | Fused throughput | Main throughput |
|---:|---:|---:|---:|---:|---:|
| 1 | 0.0369 ms | 0.0370 ms | -0.4% | 0.027M tok/s | 0.027M tok/s |
| 2 | 0.0419 ms | 0.0423 ms | -1.1% | 0.048M tok/s | 0.047M tok/s |
| 4 | 0.0531 ms | 0.0535 ms | -0.7% | 0.075M tok/s | 0.075M tok/s |
| 8 | 0.0757 ms | 0.0782 ms | -3.3% | 0.106M tok/s | 0.102M tok/s |
| 16 | 0.1024 ms | 0.1070 ms | -4.3% | 0.156M tok/s | 0.149M tok/s |
| 32 | 0.1481 ms | 0.1497 ms | -1.0% | 0.216M tok/s | 0.214M tok/s |
| 64 | 0.1979 ms | 0.1936 ms | +2.2% | 0.323M tok/s | 0.331M tok/s |
| 128 | 0.2227 ms | 0.2113 ms | +5.4% | 0.575M tok/s | 0.606M tok/s |
| 256 | 0.2512 ms | 0.2393 ms | +5.0% | 1.019M tok/s | 1.070M tok/s |
| 512 | 0.2712 ms | 0.3488 ms | -22.3% | 1.888M tok/s | 1.468M tok/s |
| 1024 | 0.3055 ms | 0.3863 ms | -20.9% | 3.351M tok/s | 2.651M tok/s |
| 2048 | 0.4039 ms | 0.4704 ms | -14.1% | 5.071M tok/s | 4.353M tok/s |
| 4096 | 0.6699 ms | 0.6859 ms | -2.3% | 6.114M tok/s | 5.972M tok/s |
| 8192 | 1.2046 ms | 1.2348 ms | -2.4% | 6.801M tok/s | 6.634M tok/s |

Fusion is neutral at the smallest shapes, loses 2-5% at 64-256 tokens where
the dense shared MLP overlaps effectively, wins 14-22% at 512-2048 tokens,
and remains 2-3% faster once both paths approach GEMM throughput limits.
Forcing overlap above the default 256-token threshold improves the main
baseline by only 0.8-4.0% at 512-8192 tokens because the FlashInfer routed MoE
and DeepGEMM shared MLP compete for the same GPU resources.

## AI assistance disclosure

AI assistance from OpenAI Codex was used to implement, debug, test, benchmark, profile, clean up, and prepare this draft PR. The human submitter is responsible for understanding and defending the change and must review every changed line and confirm the reported tests before marking the PR ready for review.

### CODE REVIEW

**Comment by xinli-sw** [comment]
related to https://github.com/vllm-project/vllm/pull/53112

**Comment by kjiang249** [comment]
Hi, thanks for working on this! Do you have an estimated timeline for merging this PR, or is it still intended to be merged?
